packages/components/src/root-container/props.ts

```diff
@@ -0,0 +1,232 @@
+  activeStateEnabled: true,
+};
+
+export function buildActiveStateProp(
+  { activeStateEnabled }: ActiveStateDomOptions,
+): PipeFunc<RootContainerProps> {
+  return buildCssProp(activeStateEnabled, DOM_CSS_CLASSES.active);
+}
+
+export function getActiveStateFromProps(
+  props: Partial<ActiveStateDomOptions> = {},
+): ActiveStateDomOptions {
+  return {
+    activeStateEnabled: props?.activeStateEnabled
+      ?? DEFAULT_ACTIVE_STATE_DOM_OPTIONS.activeStateEnabled,
+  };
+}
+
+// --- custom-attributes ---
+export type CustomAttributesDomOptions = {
+  customAttributes: AllowedDomAttributes;
+};
+
+export const DEFAULT_CUSTOM_ATTRIBUTES_DOM_OPTIONS: CustomAttributesDomOptions = {
+  customAttributes: {},
+};
+
+export function buildCustomAttributesProp(
+  { customAttributes }: CustomAttributesDomOptions,
+): PipeFunc<RootContainerProps> {
+  return buildAttributesProp(true, customAttributes);
+}
+
+export function getCustomAttributesFromProps(
+  props: Partial<CustomAttributesDomOptions> = {},
+): CustomAttributesDomOptions {
+  return {
+    customAttributes: props?.customAttributes
+      ?? DEFAULT_CUSTOM_ATTRIBUTES_DOM_OPTIONS.customAttributes,
+  };
+}
+
+// --- disabled ---
+export type DisabledDomOptions = {
+  disabled: boolean;
+};
+
+export const DEFAULT_DISABLED_DOM_OPTIONS: DisabledDomOptions = {
+  disabled: false,
+};
+
+export function buildDisabledProp(
+  { disabled }: DisabledDomOptions,
+): PipeFunc<RootContainerProps> {
+  return buildAttributesProp(disabled, {
+    [DOM_ATTRIBUTES.disabled]: ' ',
+    [DOM_ATTRIBUTES.tabIndex]: undefined,
+  });
+}
+
+export function getDisabledFromProps(
+  props: Partial<DisabledDomOptions> = {},
+): DisabledDomOptions {
+  return { disabled: props?.disabled ?? DEFAULT_DISABLED_DOM_OPTIONS.disabled };
+}
+
+// --- focus ---
+export type FocusDomOptions = {
+  focusStateEnabled: boolean;
+  tabIndex: number;
+};
+
+export const DEFAULT_FOCUS_DOM_OPTIONS: FocusDomOptions = {
+  focusStateEnabled: true,
+  tabIndex: 0,
+};
+
+export function buildFocusProp(
+  { focusStateEnabled, tabIndex }: FocusDomOptions,
+): PipeFunc<RootContainerProps> {
+  return (containerProps) => {
+    const result = buildCssProp(focusStateEnabled, DOM_CSS_CLASSES.focus)(containerProps);
+    return buildAttributesProp(
+      true,
+      {
+        [DOM_ATTRIBUTES.tabIndex]: containerProps.attributes[DOM_ATTRIBUTES.disabled]
+          ? undefined
+          : tabIndex,
+      },
+    )(result);
+  };
+}
+
+export function getFocusFromProps(
+  props: Partial<FocusDomOptions> = {},
+): FocusDomOptions {
+  return {
+    focusStateEnabled: props?.focusStateEnabled
+      ?? DEFAULT_FOCUS_DOM_OPTIONS.focusStateEnabled,
+    tabIndex: props?.tabIndex
+      ?? DEFAULT_FOCUS_DOM_OPTIONS.tabIndex,
+  };
+}
+
+// --- hint ---
+export type HintDomOptions = {
+  hint?: string;
+};
+
+export function buildHintProp(
+  { hint }: HintDomOptions,
+): PipeFunc<RootContainerProps> {
+  return buildAttributesProp(!!hint, {
```

Comment:
I set a breakpoint here and it's called many times then any prop is changed for example I change disable in the react playground

Reply:
Yes, unfortunately we've issue with cached selectors (they don't work as expected because React recreate them on each render)

I fixed it in the [separate PR](https://github.com/DevExpress/DevExtreme/pull/23809)